xw/src/main/scala/uk/ac/surrey/xw/extension/prim/OnChange.scala

```diff
@@ -0,0 +1,86 @@
+package uk.ac.surrey.xw.extension.prim
+
+import org.nlogo.api.{Argument, Context, DefaultCommand}
+import org.nlogo.api.Syntax.{StringType, commandSyntax, CommandBlockType, CommandTaskType}
+import org.nlogo.nvm.{Context => NvmContext, _}
+import org.nlogo.workspace.AbstractWorkspace
+import uk.ac.surrey.xw.api.{PropertyKey, WidgetKey}
+import uk.ac.surrey.xw.extension.{KindInfo, WidgetContextManager}
+import uk.ac.surrey.xw.state.{SetProperty => SetPropEvent, RemoveWidget, StateEvent, Writer}
+
+import scala.collection.mutable.{Publisher, Subscriber}
+import scala.collection.parallel.mutable.ParMap
+
+case class ChangeListener(func: StateEvent => Unit)  extends Subscriber[StateEvent, Publisher[StateEvent]] {
+  def notify(pub: Publisher[StateEvent], event: StateEvent): Unit = func(event)
+}
+
+object OnChange {
+  val listeners = ParMap.empty[(WidgetKey, PropertyKey), Seq[ChangeListener]]
+  def removeListeners(writer: Writer, wk: WidgetKey, pk: PropertyKey) =
+    listeners.get((wk, pk)).foreach(_.foreach(writer.removeSubscription))
+}
+
+abstract class OnChangePrim(writer: Writer, wcm: WidgetContextManager) extends DefaultCommand {
+
+  def addListener(context: Context, widgetKey: WidgetKey, propertyKey: PropertyKey, task: CommandTask): Unit = {
+    val extContext = context.asInstanceOf[ExtensionContext]
+    val ws = extContext.workspace.asInstanceOf[AbstractWorkspace]
+    OnChange.removeListeners(writer, widgetKey, propertyKey)
+
+    val listener = ChangeListener {
+      // Can run on AWT event thread, so we have explicitly submit job to JobThread. BCH 4/21/2015
+      case SetPropEvent(_, _, v, _) => runTask(ws, extContext.nvmContext, task, Array(widgetKey, v))
+      case _ =>
+    }
+
+    val removalListener: ChangeListener = ChangeListener { _ =>
+      OnChange.removeListeners(writer, widgetKey, propertyKey)
```

Comment:
My original thinking was that they need different filters, but that really doesn't make sense. In fact, maybe I should just toss the filters since I need to do pattern matching no matter what.